src/adapter/graphql/src/queries/datasets/dataset.rs

```diff
@@ -201,9 +196,78 @@ impl Dataset {
 
     /// Various endpoints for interacting with data
     async fn endpoints(&self, ctx: &Context<'_>) -> DatasetEndpoints<'_> {
-        let config = crate::utils::unsafe_from_catalog_n!(ctx, ServerUrlConfig);
+        let config = utils::unsafe_from_catalog_n!(ctx, ServerUrlConfig);
+
+        DatasetEndpoints::new(&self.dataset_request_state, config)
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug)]
+pub(crate) struct DatasetRequestState {
```

Comment:
Why do we need separate structures for r/o and mut endpoints?
Would it make sense to simplify and have just one?